pkg/storage/stores/indexshipper/compactor/deletion/request_handler.go

```diff
@@ -81,6 +90,59 @@ func (dm *DeleteRequestHandler) AddDeleteRequestHandler(w http.ResponseWriter, r
 	w.WriteHeader(http.StatusNoContent)
 }
 
+func shardDeleteRequestsByQueryRange(startTime, endTime model.Time, query, userID string, queryRange time.Duration) []DeleteRequest {
+	deleteRequests := make([]DeleteRequest, 0, endTime.Sub(startTime)/queryRange)
+	for start := startTime; start.Before(endTime); start = start.Add(queryRange) + 1 {
+		end := start.Add(queryRange)
+		if end.After(endTime) {
+			end = endTime
+		}
+
+		deleteRequests = append(deleteRequests,
+			DeleteRequest{
+				StartTime: start,
+				EndTime:   end,
+				Query:     query,
+				UserID:    userID,
+			})
+	}
+	return deleteRequests
+}
+
+func (dm *DeleteRequestHandler) queryRange(params url.Values, startTime, endTime model.Time) (time.Duration, error) {
+	qr := params.Get("max_query_range")
+	if qr == "" {
+		if dm.maxQueryRange == 0 {
+			return endTime.Sub(startTime), nil
+		}
+
+		return min(endTime.Sub(startTime), dm.maxQueryRange), nil
+	}
+
+	queryRange, err := time.ParseDuration(qr)
+	if err != nil || queryRange < time.Second {
+		return 0, errors.New("invalid query range: valid time units are 's', 'm', 'h'")
+	}
+
+	if queryRange > dm.maxQueryRange && dm.maxQueryRange != 0 {
+		dur, err := time.ParseDuration(dm.maxQueryRange.String())
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", "error parsing query range", "err", err)
+			return 0, err
+		}
```

Comment:
Good catch. Line 128 is supposed to be `model.ParseDuration` instead of `time.ParseDuration` to take advantage of the better formatting for the error message